### PR TITLE
Run tests with pegen enabled on CI

### PR DESCRIPTION
**Pull request by lysnikolaou**

### CODE REVIEW

**Comment by gvanrossum** [comment]
Hopefully Pablo can review this. :-)

**Comment by lysnikolaou** [comment]
It's test_cantrace that fails, when pegen is enabled, which is in `Lib/test/audit-tests.py`. Any idea oh what might be causing this, @pablogsal?

**Comment by pablogsal** [comment]
Checking!

**Comment by pablogsal** [comment]
Seems that the problem is that the compile() calls are not traced when the new parser is active. I am trying to investigate why

**Comment by pablogsal** [comment]
Oh, seems that there are two usages of `PySys_Audit("compile",` in `Parser/parsetok.c`!!

That seems surprising, I think these two should be moved to the builtins module.

**Comment by pablogsal** [comment]
Well, the easiest solution is add the two audit calls for now to our parser calls. The two audit calls are in:

`PyParser_ParseStringObject` and `PyParser_ParseFileObject` so we just need to add them to the equivalent functions.

**Comment by pablogsal** [comment]
I pushed a commit that adds the `PySys_Audit` calls to fix the audit tests.

**Comment by pablogsal** [comment]
There are other failures for calling `print` and `exec` in `test_grammar`. Previously, the message was:

```
>>> exec("print x")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1
    print x
          ^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(x)?
```

not the message is

```
>>> exec("print x")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<string>", line 1
SyntaxError: invalid syntax
```

**Comment by lysnikolaou** [comment]
> There are other failures for calling `print` and `exec` in `test_grammar`. Previously, the message was:
> 
> ```
> >>> exec("print x")
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "<string>", line 1
>     print x
>           ^
> SyntaxError: Missing parentheses in call to 'print'. Did you mean print(x)?
> ```
> 
> not the message is
> 
> ```
> >>> exec("print x")
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "<string>", line 1
> SyntaxError: invalid syntax
> ```

That's odd, because the tests pass. Where exactly is does this get tested?

I think this is a thing, we must generally work on improving before beta1, our error-reporting machinery. I don't think `raise_syntax_error` is enough, since there are lots of edge cases, where the error texts are different than the standard `SyntaxError: invalid syntax`.

**Comment by pablogsal** [comment]
> That's odd, because the tests pass.

Run manually `PYTHONPARSER=new ./python -m test test_grammar -v`. What do you get?

**Comment by lysnikolaou** [comment]
> > That's odd, because the tests pass.
> 
> Run manually `PYTHONPARSER=new ./python -m test test_grammar -v`. What do you get?

Multiple failures, indeed. Is there something wrong with the CI? Why doesn't `test_grammar` fail here?

**Comment by pablogsal** [comment]
Ohhhhh, this only happens when you don't pass a file:

```
~/github/pegen-cpython ci*                                                                                                                                                                                                                                                                     │
❯ ./python lel.py                                                                                                                                                                                                                                                                              │~/github/pegen-cpython ci*
  File "/home/pablogsal/github/pegen-cpython/lel.py", line 1                                                                                                                                                                                                                                   │❯ ./python -m test test_ast p
    print x                                                                                                                                                                                                                                                                                    │p
          ^                                                                                                                                                                                                                                                                                    │~/github/pegen-cpython ci*
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(x)?                                                                                                                                                                                                                    │❯ python
                                                                                                                                                                                                                                                                                               │Python 3.8.2 (default, Feb 27 2020, 01:04:54) 
~/github/pegen-cpython ci*                                                                                                                                                                                                                                                                     │[GCC 9.2.0] on linux
❯ PYTHONPARSER=new ./python lel.py                                                                                                                                                                                                                                                             │Type "help", "copyright", "credits" or "license" for more information.
  File "/home/pablogsal/github/pegen-cpython/lel.py", line 1                                                                                                                                                                                                                                   │>>> SyntaxError(1,2,3,4)
    print x                                                                                                                                                                                                                                                                                    │SyntaxError(1, 2, 3, 4)
          ^                                                                                                                                                                                                                                                                                    │>>> x = SyntaxError(1,2,3,4)
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(x)?  
```

The reason is that we set `None` here because we cannot find the text?

https://github.com/we-like-parsers/cpython/blob/pegen/Parser/pegen/pegen.c#L84

The exception machinery only injects the messages if the `text` attribute of the exception is not `None`

**Comment by pablogsal** [comment]
> Why doesn't `test_grammar` fail here?

Maybe is because the processes that `make buildbottest` spawns do not get the environment variable. Try to export it instead

**Comment by pablogsal** [comment]
> The reason is that we set `None` here because we cannot find the text?

And that seems because `p->start_rule == Py_file_input` but we are parsing a string.

**Comment by lysnikolaou** [comment]
> > The reason is that we set `None` here because we cannot find the text?
> 
> And that seems because `p->start_rule == Py_file_input` but we are parsing a string. Is that expected @lysnikolaou ?

Yeah, I introduced this bug yesterday in #35. I assumed that when compile uses something other that `Py_file_input` for `exec` and I was wrong. Sorry.

The question is what do we do about this? Can we make do with the four globals that are already there or should we add something like `enum INPUT_TYPE` back?

**Comment by pablogsal** [comment]
I found a simpler solution, which is fallback always to `get_error_line` if `p->start_rule != Py_file_input` but also if `PyErr_ProgramTextObject` failed. Pushing that commit.

**Comment by lysnikolaou** [comment]
> I found a simpler solution, which is fallback always to `get_error_line` if `p->start_rule != Py_file_input` but also if `PyErr_ProgramTextObject` failed. Pushing that commit.

Nice! How can we be confident that the environment variable is exported correctly though?

**Comment by pablogsal** [comment]
> Nice! How can we be confident that the environment variable is exported correctly though?

Temporarily, push a commit that raises or segfaults if we are using the old parser in `pythonrun`. All tests in the `build_macos_pegen` and `build_ubuntu_pegen` should pass and the others in the `build_ubuntu` and `build_macos` should fail.

**Comment by pablogsal** [comment]
@lysnikolaou Did it work as expected?

**Comment by lysnikolaou** [comment]
Nope, `test_signal` gets run many times and the tests thus hang. It probably has to do with my last change, to directly call `./python -m test ...` and not `make buildbottest ...`.

**Comment by pablogsal** [comment]
> Nope, `test_signal` gets run many times and the tests thus hang. It probably has to do with my last change, to directly call `./python -m test ...` and not `make buildbottest ...`.

Hummm, but at least the test without the old parser crash and the test with the new parser succeed? Maybe there is something in `regrtest` that does not pass environment variables....

**Comment by lysnikolaou** [comment]
> > Nope, `test_signal` gets run many times and the tests thus hang. It probably has to do with my last change, to directly call `./python -m test ...` and not `make buildbottest ...`.
> 
> Hummm, but at least the test without the old parser crash and the test with the new parser succeed? Maybe there is something in `regrtest` that does not pass environment variables....

The tests with the old parser crash immediately. In the tests with the new parser there are 11 failures.

**Comment by lysnikolaou** [comment]
Good news! I was able to verify that the tests indeed work as expected. Only one question remaining. Now that I have introduced a bug when invoking `python` with the old parser (a `SystemError` gets thrown), some of the tests that spawn a new `python` process by calling `subprocess.Popen` do not inherit the environment variable. Is that expected?

**Comment by lysnikolaou** [comment]
Something is off here. When I run python locally, I get the following output:

With the old parser:
```
╰─ ./python      
Python 3.9.0a5+ (heads/ci:849809ac79, Apr  5 2020, 01:28:19) 
[GCC 9.2.1 20191008] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> exec('1+1')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
SystemError: temporary
```

With the new parser:
```
╰─ ./python -pnew                           
Python 3.9.0a5+ (heads/ci:849809ac79, Apr  5 2020, 01:28:19) 
[GCC 9.2.1 20191008] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> exec
<built-in function exec>
>>> exec('1+1')
>>> exec('1+1')
>>> exit()
```

But on the Pegen tests, there are calls to `exec` that fail with `SystemError`, although the line where this gets thrown should never be reached.

**Comment by lysnikolaou** [comment]
To make things even more interesting, when I run `./python -m test -v test_grammar` everything runs smoothly. No errors. 

BUT(!) when I run `./python -m test -j2 -v test_grammar` all hell breaks loose and I get multiple failures with `SystemError`.

**Comment by lysnikolaou** [comment]
It seems that invoking `./python -m test` with the `j<something>` flag also uses `subprocess.Popen`, so it all comes down to this one question:

Shouldn't the child process, spawned by `subprocess.Popen`, inherit the environment, since no `env` kwarg is passed to it?

**Comment by lysnikolaou** [comment]
I guess, it was just short-sightedness on my part. I just had to comment out the `-E` flag in `run_tests.py`. I'll have a look at the test results, which should show what we need to work on and start opening issues tomorrow.

Sorry for the noise.

**Comment by pablogsal** [comment]
I gave a quick look at some of the errors and seems that there is a bunch of cases when we are throwing generic syntax errors and the old parser give specific reasons. For instance:

Old parser:
```
>>> f(x=3, 4)
  File "<stdin>", line 1
SyntaxError: positional argument follows keyword argument
```

new parser:

```
>>> f(x=1, y)                                                                                                                                                                                                                                                                                                                                                                                  
  File "<stdin>", line 1                                                                                                                                                                                                                                                                                                                                                                       
    f(x=1, y)                                                                                                                                                                                                                                                                                                                                                                                  
            ^                                                                                                                                                                                                                                                                                                                                                                                  
SyntaxError: invalid syntax    
```

A way of fixing this could be allowing that alternative to be parser and then fail:

```diff
@ Grammar/python.gram:440 @ args[expr_ty]:
                     : CHECK(singleton_seq(p, a)),
                 (b) ? ((expr_ty) b)->v.Call.keywords : NULL,
                 EXTRA) }
    | a=kwargs ',' args { RAISE_SYNTAX_ERROR(p, "oh no!") }
    | a=kwargs { _Py_Call(CONSTRUCTOR(p),
                          CHECK_NULL_ALLOWED(seq_extract_starred_exprs(p, a)),
                          CHECK_NULL_ALLOWED(seq_delete_starred_exprs(p, a)),

```

with that change:

```
>>> f(x=3, 4)
  File "<stdin>", line 1
    f(x=3, 4)
            ^
SyntaxError: oh no!
```

Not sure if this is the best way, but is certainly a way.

**Comment by gvanrossum** [comment]
Sigh. We may have to track down every specialized error message in the old parser that's happening in ast.c and emulate it. (Fortunately errors detected in bytecode generation are unaffected, like `'return' outside function` or `'break' outside loop`.)

**Comment by lysnikolaou** [comment]
I'd say we land this and work on fixing the tests in new PRs.

File: Tools/scripts/run_tests.py [hunk 1]
```diff
@@ -25,7 +25,7 @@ def main(regrtest_args):
             '-u',                 # Unbuffered stdout and stderr
             '-W', 'default',      # Warnings set to 'default'
             '-bb',                # Warnings about bytes/bytearray
-            '-E',                 # Ignore environment variables
+            #'-E',                 # Ignore environment variables
```
Comment (pablogsal) [inline Tools/scripts/run_tests.py:28]: Instead of removing -E, could we pass down the flag (-p new) if the parent has the flag or sees the environment variable?

Alternatively, we could just remove -E if the parent sees the environment variable, but in general I would recommend against just remove -E all the time as this was there to prevent errors due to environment interference in tests.

**Comment by pablogsal** [comment]
Seems that all test are passing with e2096d7 so something is wrong....

**Comment by lysnikolaou** [comment]
> Seems that all test are passing with [e2096d7](https://github.com/we-like-parsers/cpython/commit/e2096d78ce35a55a8385922e9d6ea261d4435fdd) so something is wrong....

I'm checking.

**Comment by lysnikolaou** [comment]
This seems to work fine now.

**Comment by pablogsal** [comment]
> I'd say we land this and work on fixing the tests in new PRs.

Yeah I agree, let's land this and start fixing individual errors
